### PR TITLE
ASoC: SOF: skl add cl_stream_prepare_skl

### PR DESCRIPTION
**Pull request by zhuyingjiang**
add the cl_stream_prepare_skl to alloc the DMA buffer and init the DMA

### CODE REVIEW

File: sound/soc/sof/intel/hda-loader.c [hunk 1]
```diff
@@ -372,3 +372,156 @@ int hda_dsp_cl_boot_firmware(struct snd_sof_dev *sdev)
...
+		ret = hda_dsp_core_reset_power_down(sdev, HDA_DSP_CORE_MASK(0));
+		if (ret < 0) {
+			dev_err(sdev->dev, "dsp core0 disable fail: %d\n", ret);
+			return ret;
+		}
+		ret = hda_dsp_enable_core(sdev, HDA_DSP_CORE_MASK(0));
+	}
+
+	/* prepare DMA for code loader stream */
+	ret = cl_stream_prepare_skl(sdev, 0x40, fwsize, &sdev->dmab,
```
Comment (lgirdwood) [inline sound/soc/sof/intel/hda-loader.c]: where does 0x40 come from ? this should probably be a macro

File: sound/soc/sof/intel/hda-loader.c [hunk 2]
```diff
@@ -34,6 +35,8 @@
 #include "../ops.h"
 #include "hda.h"
 
+#include "../../intel/skylake/skl-sst-cldma.h"
```
Comment (lgirdwood) [inline sound/soc/sof/intel/hda-loader.c:38]: If we need any macros we should define locally.

File: sound/soc/sof/intel/hda-loader.c [hunk 3]
```diff
@@ -373,26 +376,147 @@ int hda_dsp_cl_boot_firmware(struct snd_sof_dev *sdev)
 	return ret;
 }
 
+/* Code loader helper APIs */
+static int cl_skl_cldma_setup_bdle(struct snd_sof_dev *sdev,
```
Comment (lgirdwood) [inline sound/soc/sof/intel/hda-loader.c]: We already have a function for creating BDLE and should reuse it.

File: sound/soc/sof/intel/hda-loader.c [hunk 4]
```diff
@@ -373,26 +376,147 @@ int hda_dsp_cl_boot_firmware(struct snd_sof_dev *sdev)
...
+ * Setup controller
+ * Configure the registers to update the dma buffer address and
+ * enable interrupts.
+ * Note: Using the channel 1 for transfer
+ */
+static void cl_skl_cldma_setup_controller(struct snd_sof_dev *sdev,
+		struct snd_dma_buffer *dmab_bdl, unsigned int max_size,
+		u32 count)
+{
+	/* make sure Run bit is cleared before setting stream register */
```
Comment (lgirdwood) [inline sound/soc/sof/intel/hda-loader.c:443]: ok, so I guess this is where we differ from APL, if so is it all the register write below or only some ?

File: sound/soc/sof/intel/hda-loader.c [hunk 5]
```diff
@@ -373,26 +376,147 @@ int hda_dsp_cl_boot_firmware(struct snd_sof_dev *sdev)
...
+	cl_skl_cldma_stream_run(sdev, 0);
+
+	snd_sof_dsp_update_bits(sdev, HDA_DSP_HDA_BAR, SKL_ADSP_REG_CL_SD_CTL,
+				CL_SD_CTL_IOCE_MASK, CL_SD_CTL_IOCE(0));
+	snd_sof_dsp_update_bits(sdev, HDA_DSP_HDA_BAR, SKL_ADSP_REG_CL_SD_CTL,
+				CL_SD_CTL_FEIE_MASK, CL_SD_CTL_FEIE(0));
+	snd_sof_dsp_update_bits(sdev, HDA_DSP_HDA_BAR, SKL_ADSP_REG_CL_SD_CTL,
+				CL_SD_CTL_DEIE_MASK, CL_SD_CTL_DEIE(0));
+	snd_sof_dsp_update_bits(sdev, HDA_DSP_HDA_BAR, SKL_ADSP_REG_CL_SD_CTL,
```
Comment (lgirdwood) [inline sound/soc/sof/intel/hda-loader.c:452]: These 4 writes are to same register, cant they be done as one write ?

File: sound/soc/sof/intel/hda-loader.c [hunk 6]
```diff
@@ -373,26 +376,147 @@ int hda_dsp_cl_boot_firmware(struct snd_sof_dev *sdev)
...
+			  count - 1);
+	snd_sof_dsp_update_bits(sdev, HDA_DSP_HDA_BAR, SKL_ADSP_REG_CL_SD_CTL,
+				CL_SD_CTL_IOCE_MASK, CL_SD_CTL_IOCE(1));
+	snd_sof_dsp_update_bits(sdev, HDA_DSP_HDA_BAR, SKL_ADSP_REG_CL_SD_CTL,
+				CL_SD_CTL_FEIE_MASK, CL_SD_CTL_FEIE(1));
+	snd_sof_dsp_update_bits(sdev, HDA_DSP_HDA_BAR, SKL_ADSP_REG_CL_SD_CTL,
+				CL_SD_CTL_DEIE_MASK, CL_SD_CTL_DEIE(1));
+	snd_sof_dsp_update_bits(sdev, HDA_DSP_HDA_BAR, SKL_ADSP_REG_CL_SD_CTL,
+				CL_SD_CTL_STRM_MASK,
+				CL_SD_CTL_STRM(FW_CL_STREAM_NUMBER));
```
Comment (lgirdwood) [inline sound/soc/sof/intel/hda-loader.c:480]: ditto, we also need comments here.

File: sound/soc/sof/sof-priv.h [hunk 7]
```diff
@@ -327,6 +327,8 @@ struct snd_sof_dev {
 	/* firmware loader */
 	int cl_bar;
 	struct snd_dma_buffer dmab;
+	struct snd_dma_buffer dmab_bdl;
```
Comment (lgirdwood) [inline sound/soc/sof/sof-priv.h:330]: do we really need this since we already have some existing bdl logic

**Comment by zhuyingjiang** [comment]
Sorry, I use the error branch to push another PR, and now I update it.
WIth this PR 2310452bd36b is to enable the core, and I think it is OK to merge now with some comments added.
The second is b0c3460e946bc and is only to change the old ops from apl to skl, this is also ok to merge
The third is to alloc the dma buffer, 81b2f482f850b1 , I will do change with your comments.

File: sound/soc/sof/intel/hda-loader.c (hunk 2 above)
Comment (zhuyingjiang) [reply sound/soc/sof/intel/hda-loader.c:38]: OK, my question is to create a new head file like hda-skl.h, or just define it in hda-loader.c now? For there are too much defines for the registers, I prefer using a new head file..

File: sound/soc/sof/intel/hda-loader.c (hunk 3 above)
Comment (zhuyingjiang) [reply sound/soc/sof/intel/hda-loader.c]: OK, I will check if it can be reused.

**Comment by zhuyingjiang** [comment]
@lgirdwood

File: sound/soc/sof/intel/hda-loader.c (hunk 2 above)
Comment (lgirdwood) [reply sound/soc/sof/intel/hda-loader.c:38]: The registers should mostly be the same as existing ones for APL, you may need to check each one being used.

**Pull request closed without merging**